### ISSUE TITLE
Proposal: Allow Multiple ContentDialogs to be Open

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#1679]
# Proposal: Allow Multiple ContentDialogs to be Open

As stated, allow multiple content dialogs to be displayed on top of each other (like classic Windows allows) and make it the responsibility of the App to manage them correctly.

## Summary

There are some cases where multiple content dialogs are almost unavoidable (See rationale below). However, if you try to show more than one the app crashes -- hard. 

There are two ways to fix this without the runtime closing a content dialog automatically (which is not OK):
 1. Allow multiple content dialogs to just render on top of each other. It's up to the app to manage them correctly.
 2. Provide a rock-solid way of determining when a content dialog is open and then a way of closing it before showing the next one.

## Rationale

Currently, only one content dialog can be open at a time in the UI. This makes sense as it's meant to be blocking waiting for user input.

However, there are some scenarios where notifications are being sent from the backend to the frontend (not as a result of direct user input) and they should be displayed in a blocking content dialog to the user.

The problem occurs when multiple notifications are sent from the backend and the user hasn't yet closed the first one. In this case the app crashes when it tries to create the second ContentDialog. As a work around, apps can attempt to close dialogs themselves; however, this does not always work. If a request to open a content dialog comes fast enough a crash still occurs.

Applications should not crash so easily when opening a ContentDialog. Work-arounds such as the example code below are not reliable as they are somewhat dependent on timing.

```csharp
    public static bool IsContentDialogOpen(Window window = null)
        {
            // By default use the current window
            if (window == null)
            {
                window = Window.Current;
            }

            // Get all open popups in the window. A ContentDialog is a popup.
            var popups = VisualTreeHelper.GetOpenPopups(window);

            foreach (var popup in popups)
            {
               if(popup.Child is ContentDialog)
               {
                  // A content dialog is open
                  return (true);
               }
            }

            return (false);
        }
```

or

```csharp
var popups = VisualTreeHelper.GetOpenPopups(Window.Current);
foreach(var popup in popups)
{
    if (popup.Child is ContentDialog)
    {
        ((ContentDialog)popup.Child).Hide();
    }
}
```

**Comment by mrlacey** [comment]
That trying to open multiple dialogs causes a crash is a bug and should be treated as such.

I'm not convinced by the rationale for adding this.
- How does an app indicate the context of each dialog when multiple ones are shown? I'd expect having multiple dialogs open at once to be potentially confusing to the user and the developer/designer would likely need to do a lot of work to make sure this is always clear to all users. I can easily see the amount of work needed for this as being much more than ensuring that the app doesn't try and open a dialog when another is already open.
- Having the "backend" open dialogs seems wrong. You risk all sorts of potential issues when the backend is directly controlling the UI in this way due to not separating the responsibilities. The "backend" also becomes tightly coupled to the UI layer. At best I would expect the backend to tell the frontend that it needs some information and the frontend is responsible for requesting and returning that information. This leaves the frontend responsible for not prompting for multiple things at the same time.

**Comment by yaichenbaum** [comment]
I have a Notepad app that has a content dialog with a text box that lets the user goto a different line on the main editing text box. If that dialog is open and the user presses the close button on the app it will try to open a confirm dialog if there are unsaved changes. The second that second dialog comes up the app will crash because there is already an existing content dialog open.
In my case I need to close any open dialogs which as mentioned doesn't always work. I don't know if allowing more then one content dialog to be open at a time is a good idea, but there should definitely be a reliable way to close existing content dialogs so another one can be shown.

**Comment by robloo** [comment]
@mrlacey We could talk about application architecture endlessly but that would be a bit out of scope I think. However, since you are curious here is the full scenario:

 * User starts a search which runs in the backend (Here it is possible to press ‘Enter’ very rapidly in the frontend and schedule two searches one after the other. Yes, that could be changed/filtered but complexity goes up a lot compared to just allowing it.)
 * The backend completes the first search but finds no results. It notifies the frontend that ‘Search is complete with no results’
 * The frontend displays the confirmation ContentDialog ‘Search is complete with no results’. The user does nothing (and has no chance to due to the speed)
 * The backend completes the second search but finds no results. It again notifies the frontend that ‘Search is complete with no results’
 * The frontent detects the first ContentDialog is open and attempts to close it. However, a lot of the WinUI code is not syncronous and there are some timing issues I cannot control. It's possible without any async code on my part that the second ContentDialog is opened before the first has a chance to fully close. (I would have to wait for some ridiculous long time to be safe which is just bad.)
 * In the majority of cases the first ContentDialog is closed successfully by my applicaiton using the example code above. However, in some cases -- just depending on timing -- the app will crash.

So I try to close any already open ContentDialogs but the APIs to do so just are not there. Eventually the app will crash if requests are coming fast enough. I do of course agree that showing multiple ContentDialogs is bad but there are some edge cases as @yaichenbaum also mentions.

The point is we can’t adequately manage ContentDialogs because the API just isn't there:
 1. There is no way to reliably detects when a ContentDialog is open
 2. There is no way to reliably close the open ContentDialog (synchronously before the next opens)
 3. There is also no parameter in ContentDialgo.ShowAsync() specifying what should happen to any existing ContentDialogs.

You can't have the platform hard-crash when a second ContentDialog is being opened and then not provide an API to manage ContentDialogs (instead relying on the user to do it). On top of that – the limit of one ContentDialog open should not be enforced by crashing… classic Windows or WPF never did that. The framework should allow multiple ContentDialogs to be open for the cases where the app doesn’t manage them correctly (assuming they get the API to do so). 

Crashes are bad, ContentDialogs should not be this fragile.

I'm going to reword my proposal a bit asking for the API to manage ContentDialogs and also for the app not to crash when a second is opened. I think we need both to robustly fix this for application developers.

**Comment by eugenegff** [comment]
User attention is a resource, and as always, exclusive access to resource should be queued. 

That is especially true when current user activity contains asynchronous continuations, like in chain like 'Ctrl + S shortcut' => FileSaveDialog.ShowAsync => StorageFile.WriteAsync => 'HideModifiedMark or ShowError'. Any background activity should queue attempt to access user till the current and all previous activities would be completed.

**Comment by robloo** [comment]
@eugenegff I agree with everything you said and that is certainly the best way of handling it. Implementing a queuing system would be a lot easier if there was an 'IsContentDialogOpen' API some place in the platform though. Otherwise developers have to write extra code to manage this themselves.

The app also should be able to make a judgement call of whether to close an existing ContentDialog and show another without waiting for user input. The API's to do that cleanly don't exist.

And finally, I still feel as a fail-safe the platform should not crash if a second ContentDialog is opened.

**Comment by jevansaks** [comment]
We agree that multiple ContentDialogs should not cause a crash. There's additionally a bug in the platform today with XAML Islands where each island can't have a dialog open because the tracking of this state is per-thread.

**Comment by zhuxb711** [comment]
Here is my solution for multiple ContentDialog, hope it would help you @robloo  :

```csharp
    public class QueueContentDialog : ContentDialog
    {
        private static readonly AutoResetEvent Locker = new AutoResetEvent(true);

        private static int WaitCount = 0;

        public static bool IsRunningOrWaiting
        {
            get
            {
                return WaitCount != 0;
            }
        }

        private bool IsCloseRequested = false;
        private ContentDialogResult CloseWithResult;

        public new async Task<ContentDialogResult> ShowAsync()
        {
            _ = Interlocked.Increment(ref WaitCount);

            await Task.Run(() =>
            {
                Locker.WaitOne();
            });

            var Result = await base.ShowAsync();

            _ = Interlocked.Decrement(ref WaitCount);

            Locker.Set();

            if (IsCloseRequested)
            {
                IsCloseRequested = false;
                return CloseWithResult;
            }
            else
            {
                return Result;
            }
        }

        public void Close(ContentDialogResult CloseWithResult)
        {
            IsCloseRequested = true;
            this.CloseWithResult = CloseWithResult;
            Hide();
        }

        public QueueContentDialog()
        {
            if (AppThemeController.Current.Theme == ElementTheme.Dark)
            {
                Background = Application.Current.Resources["DialogAcrylicBrush"] as Brush;
                RequestedTheme = ElementTheme.Dark;
            }
            else
            {
                RequestedTheme = ElementTheme.Light;
            }
        }
    }
```

**Comment by lukasf** [comment]
I have also run into this trouble. Sure this can be handled in the app somehow, but it is tedious work that must be repeated in every single app. A framework solution would be highly appreciated!

I like the proposal, but I have some remarks:

1. This also affects MessageDialog! The solution must handle both ContentDialog and MessageDialog.

2. The `CloseExisting` enum value should close the existing dialog with `ContentDialogResult.None` (MessageDialog would return `null`).

3. Please also consider an enum value `Enqueue`. Then the dialog would be enqueued and shown once all previous dialogs are closed. The method is async anyways, so we just have to wait a little longer. I think this should be the default!

4. If people really want multiple dialogs at the same time, this could be a fourth option. But I think it should not be the default, because it is a pretty irritating user experience (and maybe this should not be implemented at all).

Scenarios such as @yaichenbaum's could perhaps be better implemented by a control that is placed in the normal view (on top level, hovering the normal content), in a Popup, or placed in a specific search area that opens inside normal app layout.

**Comment by robloo** [comment]
@zhuxb711 Thanks for the code. Tracking open content dialogs is totally possible to do on the app-side. However, it's added work that shouldn't be required. This also gets more complex to coordinate as apps become larger/teams grow. In the older frameworks we never had to worry about such things and could just show dialogs at will without crashes. It would be ideal to have a framework solution in WinUI.

@lukasf 

1. True! MessageDialog is right now part of the OS but with WinUI 3.0 it can be fixed as well. I've also moved on to ContentDialog exclusively in all my code as MessageDialog seems to be much less preferred to use. I'll add some comments including MessageDialog to the description.

2. Yes, I stated 'default button press' but had I thought about it even 2 seconds longer you are of course correct.

3. Good idea, I'll add it! Hate the enum name I came up with but, well, it's a work-in-progress.

4. I don't think multiple dialogs are a good thing at all. My point was it's up to the app to prevent it not the platform. This will align with WPF for people bringing over old apps too. Certainly the platform shouldn't crash on trying to open the second dialog!

**Comment by JasonStein** [comment]
This is easily achievable on user end including the "queuing logic", here is the class my friend and I wrote in Notepads to handle two scenarios perfectly (you can choose to wait on previous one or not), it has been running in production for half a year and I am never seeing exception again:

https://github.com/JasonStein/Notepads/blob/f127d170c16cbf0831c2cddb480a3ea05e202930/src/Notepads/Utilities/DialogManager.cs

```cs
OpenDialogAsync(NotepadsDialog dialog, bool awaitPreviousDialog)
```

```cs
    public class NotepadsDialog : ContentDialog
    {
        public bool IsAborted = false;
    }
```
"IsAborted" here is to keep track if current dialog is cancelled and aborted by DialogManager, so you are fully aware of the context. 

```cs
public static class DialogManager
    {
        public static NotepadsDialog ActiveDialog;

        private static TaskCompletionSource<bool> _dialogAwaiter = new TaskCompletionSource<bool>();

        public static async Task<ContentDialogResult> OpenDialogAsync(NotepadsDialog dialog, bool awaitPreviousDialog)
        {
            return await OpenDialog(dialog, awaitPreviousDialog);
        }

        static async Task<ContentDialogResult> OpenDialog(NotepadsDialog dialog, bool awaitPreviousDialog)
        {
            TaskCompletionSource<bool> currentAwaiter = _dialogAwaiter;
            TaskCompletionSource<bool> nextAwaiter = new TaskCompletionSource<bool>();
            _dialogAwaiter = nextAwaiter;

            if (ActiveDialog != null)
            {
                if (awaitPreviousDialog)
                {
                    await currentAwaiter.Task;
                }
                else
                {
                    ActiveDialog.IsAborted = true;
                    ActiveDialog.Hide();
                }
            }

            ActiveDialog = dialog;
            var result = await ActiveDialog.ShowAsync();
            nextAwaiter.SetResult(true);
            return result;
        }
    }
```

**Comment by mdtauk** [comment]
I think what the request here is something between a single, one at a time, ContentDialog - and not allowing an app to spam the window with endless dialogs, preventing the user from feeling in control of the experience.

A queue sounds like a sensible idea.  One dialog visible at a time, but the closing of it, will spawn the next.

Perhaps there should be a limit on the number of spaces in a queue?  Maybe there could be a behaviour to allow more than say 4 at a time.

For scenarios with multiple documents displaying confirmation dialogs on closure - there could be guidance to combine those into a single dialog maybe.

**Comment by chingucoding** [comment]
I totally agree with @mdtauk here, using a queue seems to be the best choice UX wise (in my opinion). Having multiple ContentDialogs open at the same time can end up just being confusing.

Also a limitation to the maximum amount of ContentDialogs seems very reasonable, as to many dialogs might overwhelm the user.

**Comment by eugenegff** [comment]
> I totally agree with @mdtauk here, using a queue seems to be the best choice UX wise (in my opinion). Having multiple ContentDialogs open at the same time can end up just being confusing.
> 
> Also a limitation to the maximum amount of ContentDialogs seems very reasonable, as to many dialogs might overwhelm the user.

That is so only till you think about multi window situation. Having one ContentDialog per window is perfectly reasonable, and not being able to continue work in one window due to the not closed content dialog in some other window - is not acceptable

**Comment by mdtauk** [comment]
In UWP there is AppWindow which shares the UI Thread, so if ContentDialog is thread based, then that would indeed cause problems for all windows the app has created.

In WinUI Desktop / Win32 each window has it's own thread, so these queues would be per thread/per window.

UWP would be limited unless these queues are not per thread, but per Window - and were not blocking of other windows.

**Comment by eugenegff** [comment]
In Win32 you can have several HWND windows in single thread, as well as non modal dialogs.
https://docs.microsoft.com/en-us/windows/win32/dlgbox/using-dialog-boxes#creating-a-modeless-dialog-box

**Comment by JasonStein** [comment]
> I think what the request here is something between a single, one at a time, ContentDialog - and not allowing an app to spam the window with endless dialogs, preventing the user from feeling in control of the experience.
> 
> A queue sounds like a sensible idea. One dialog visible at a time, but the closing of it, will spawn the next.
> 
> Perhaps there should be a limit on the number of spaces in a queue?
> Maybe there could be a permission to allow more than say 4 at a time?
> 
> For scenarios with multiple documents displaying confirmation dialogs on closure - there could be guidance to combine those into a single dialog maybe.

You don't need to create a queue just for this and maintaining the life cycle of the queue is error prone in terms of app logic and will make users confuse. Dev should just call OpenDialog whenever he or she want at the right moment instead of putting it in a queue and make things complicated. And this is what I am doing for Notepads. The solution I have worked perfectly and I do not need to care about when l can or what should I do before calling. Here are some screen records:

Wait scenario: 
![ezgif-2-dd3e6fef723b](https://user-images.githubusercontent.com/1166162/77459244-a8b5fe00-6dbc-11ea-90a3-5440230ec0cf.gif)

Overwrite scenarios: 
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/1166162/77459928-a607d880-6dbd-11ea-9f41-8d9ab0411932.gif)

**Comment by mrlacey** [comment]
This should probably also x-Ref https://github.com/microsoft/microsoft-ui-xaml/issues/885

**Comment by robloo** [comment]
Perhaps still valid

**Comment by castorix** [comment]
> There is still no reliable way to ensure a ContentDialog is closed before opening the next one.

To avoid "_Only a single ContentDialog can be open at any time_" with not too much code
I use a boolean in ContentDialog.Closed,
but it should be handled internally by ContentDialog...

**Comment by Balkoth** [comment]
If it would be this easy once there are multiple tasks involved and the `ContentDialog` is used as a progress dialog, things get really hairy...

`ContentDialog.Hide()` is not sufficient to ensure the dialog is closed! I attach to the `OnClosed` event for now and only after waiting for a boolean `IsClosed` is set i try to show a dialog again.

**Comment by zhuxb711** [comment]
@Balkoth I would share a better implementation which is lock free:

```csharp
    internal readonly struct QueueContentDialogCoreData
    {
        public ContentDialog Instance { get; }

        public SynchronizationContext SyncContext { get; }

        public TaskCompletionSource<ContentDialogResult> TaskSource { get; } = new TaskCompletionSource<ContentDialogResult>();

        public QueueContentDialogCoreData(QueueContentDialog Instance, SynchronizationContext SyncContext)
        {
            this.Instance = Instance;
            this.SyncContext = SyncContext;
        }
    }

    internal class QueueContentDialog : ContentDialog
    {
        private static bool InternalShowingDialogFlag;
        private static readonly BlockingCollection<QueueContentDialogCoreData> DialogCollection = new BlockingCollection<QueueContentDialogCoreData>();
        private static readonly Thread BackgroundProcessThread = new Thread(ProcessThread)
        {
            IsBackground = true,
            Priority = ThreadPriority.BelowNormal
        };

        public static bool HaveAnyPresents => DialogCollection.Count > 0 || InternalShowingDialogFlag;

        public bool IsOpened { get; private set; }

        private static void ProcessThread()
        {
            foreach (QueueContentDialogCoreData Data in DialogCollection.GetConsumingEnumerable())
            {
                InternalShowingDialogFlag = true;

                try
                {
                    Data.TaskSource.TrySetResult(Data.SyncContext.PostAsync(() => Data.Instance.ShowAsync().AsTask()).WaitAndUnwrapException());
                }
                catch (Exception ex)
                {
                    Data.TaskSource.TrySetException(ex);
                }
                finally
                {
                    InternalShowingDialogFlag = false;
                }
            }
        }

        public new Task<ContentDialogResult> ShowAsync()
        {
            QueueContentDialogCoreData Data = new QueueContentDialogCoreData(this, SynchronizationContext.Current);
            DialogCollection.Add(Data);
            return Data.TaskSource.Task;
        }

        static QueueContentDialog()
        {
            BackgroundProcessThread.Start();
        }

        protected QueueContentDialog(XamlRoot Root)
        {
            IsTabStop = false;
            XamlRoot = Root;
            DefaultButton = ContentDialogButton.Primary;
            Opened += QueueContentDialog_Opened;
            Closed += QueueContentDialog_Closed;
        }

        private void QueueContentDialog_Closed(ContentDialog sender, ContentDialogClosedEventArgs args)
        {
            IsOpened = false;
        }

        private void QueueContentDialog_Opened(ContentDialog sender, ContentDialogOpenedEventArgs args)
        {
            IsOpened = true;
        }
    }
```
